File: application/backend-credit-card-enrollment/backend-c#/Common/Projection/ProjectionController.cs
using CreditCardEnrollment.Common.Ambar;
using CreditCardEnrollment.Common.SerializedEvent;
using MongoDB.Bson;

namespace CreditCardEnrollment.Common.Projection;

public abstract class ProjectionController {
    private readonly MongoTransactionalProjectionOperator _mongoOperator;
    private readonly Deserializer _deserializer;
    private readonly ILogger<ProjectionController> _logger;

    protected ProjectionController(
        MongoTransactionalProjectionOperator mongoOperator, 
        Deserializer deserializer,
        ILogger<ProjectionController> logger
    ) {
        _mongoOperator = mongoOperator;
        _deserializer = deserializer;
        _logger = logger;
    }

    protected string ProcessProjectionHttpRequest(
        AmbarHttpRequest ambarHttpRequest,
        ProjectionHandler projectionHandler,
        string projectionName) {
        try {
            _logger.LogDebug(
                "Starting to process projection for event name: {EventName} using handler: {HandlerName}", 
                ambarHttpRequest.SerializedEvent.EventName,
                projectionHandler.GetType().Name
            );
            var @event = _deserializer.Deserialize(ambarHttpRequest.SerializedEvent);

            _mongoOperator.StartTransaction();
            var isAlreadyProjected = _mongoOperator
                .CountDocuments<BsonDocument>(
                    "ProjectionIdempotency_ProjectedEvent", 
                    doc => doc["eventId"] == @event.EventId && 
                           doc["projectionName"] == projectionName
                ) != 0;

            if (isAlreadyProjected) {
                _mongoOperator.AbortDanglingTransactionsAndReturnSessionToPool();
                _logger.LogDebug(
                    "Duplication projection ignored for event name: {EventName} using handler: {HandlerName}", 
                    ambarHttpRequest.SerializedEvent.EventName,
                    projectionHandler.GetType().Name
                );
                return AmbarResponseFactory.SuccessResponse();
            }

            var projectedEvent = new BsonDocument
            {
                { "eventId", @event.EventId },
                { "projectionName", projectionName }
            };

            _mongoOperator
                .InsertOne("ProjectionIdempotency_ProjectedEvent", projectedEvent);

            projectionHandler.Project(@event);

            _mongoOperator.CommitTransaction();
            _mongoOperator.AbortDanglingTransactionsAndReturnSessionToPool();

            _logger.LogDebug(
                "Projection successfully processed for event name: {EventName} using handler: {HandlerName}", 
                ambarHttpRequest.SerializedEvent.EventName,
                projectionHandler.GetType().Name
            );
            return AmbarResponseFactory.SuccessResponse();
        } catch (Exception ex) when (ex.Message?.StartsWith("Unknown event type") == true) {
            _mongoOperator.AbortDanglingTransactionsAndReturnSessionToPool();
            
            _logger.LogDebug(
                "Unknown event in projection ignored for event name: {EventName} using handler: {HandlerName}", 
                ambarHttpRequest.SerializedEvent.EventName,
                projectionHandler.GetType().Name
            );
            return AmbarResponseFactory.SuccessResponse();
        } catch (Exception ex) {
            _mongoOperator.AbortDanglingTransactionsAndReturnSessionToPool();
            _logger.LogError(
                "Exception in ProcessProjectionHttpRequest: {0}, {1}. For event name: {EventName} using handler: {HandlerName}", 
                ex.Message, 
                ex.StackTrace,
                ambarHttpRequest.SerializedEvent.EventName,
                projectionHandler.GetType().Name
            );
            return AmbarResponseFactory.RetryResponse(ex);
        }
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/common/projection/ProjectionController.ts
import { MongoTransactionalProjectionOperator } from './MongoTransactionalProjectionOperator';
import { Deserializer } from '../serializedEvent/Deserializer';
import { AmbarHttpRequest } from '../ambar/AmbarHttpRequest';
import { AmbarResponseFactory } from '../ambar/AmbarResponseFactory';
import { ProjectionHandler } from './ProjectionHandler';
import { log } from '../util/Logger';

export abstract class ProjectionController {
    protected constructor(
        private readonly mongoOperator: MongoTransactionalProjectionOperator,
        private readonly deserializer: Deserializer,
    ) {}

    protected async processProjectionHttpRequest(
        ambarHttpRequest: AmbarHttpRequest,
        projectionHandler: ProjectionHandler,
        projectionName: string
    ): Promise<string> {
        try {
            log.debug(
                `Starting to process projection for event name: ${ambarHttpRequest.payload.event_name} using handler: ${projectionHandler.constructor.name}`
            );

            const event = this.deserializer.deserialize(ambarHttpRequest.payload);

            await this.mongoOperator.startTransaction();

            const isAlreadyProjected = await this.mongoOperator.countDocuments(
                'ProjectionIdempotency_ProjectedEvent',
                {
                    eventId: event.eventId,
                    projectionName: projectionName
                }
            ) !== 0;

            if (isAlreadyProjected) {
                await this.mongoOperator.abortDanglingTransactionsAndReturnSessionToPool();
                log.debug(
                    `Duplication projection ignored for event name: ${ambarHttpRequest.payload.event_name} using handler: ${projectionHandler.constructor.name}`
                );
                return AmbarResponseFactory.successResponse();
            }

            // Record projected event
            await this.mongoOperator.insertOne('ProjectionIdempotency_ProjectedEvent', {
                eventId: event.eventId,
                projectionName: projectionName
            });

            await projectionHandler.project(event);

            await this.mongoOperator.commitTransaction();
            await this.mongoOperator.abortDanglingTransactionsAndReturnSessionToPool();

            log.debug(
                `Projection successfully processed for event name: ${ambarHttpRequest.payload.event_name} using handler: ${projectionHandler.constructor.name}`
            );
            return AmbarResponseFactory.successResponse();

        } catch (ex) {
            if (ex instanceof Error && ex.message.startsWith('Unknown event type')) {
                await this.mongoOperator.abortDanglingTransactionsAndReturnSessionToPool();

                log.debug(
                    `Unknown event in projection ignored for event name: ${ambarHttpRequest.payload.event_name} using handler: ${projectionHandler.constructor.name}`
                );
                return AmbarResponseFactory.successResponse();
            }

            await this.mongoOperator.abortDanglingTransactionsAndReturnSessionToPool();
            log.error(
                `Exception in ProcessProjectionHttpRequest: ${ex}. For event name: ${ambarHttpRequest.payload.event_name} using handler: ${projectionHandler.constructor.name}`
            );
            return AmbarResponseFactory.retryResponse(ex as Error);
        }
    }
}